include/kernel.h

```diff
@@ -3079,199 +3088,80 @@ static inline u32_t k_mem_slab_num_free_get(struct k_mem_slab *slab)
  * @cond INTERNAL_HIDDEN
  */
 
-/*
- * Memory pool requires a buffer and two arrays of structures for the
- * memory block accounting:
- * A set of arrays of k_mem_pool_quad_block structures where each keeps a
- * status of four blocks of memory.
- */
-struct k_mem_pool_quad_block {
-	char *mem_blocks; /* pointer to the first of four memory blocks */
-	u32_t mem_status; /* four bits. If bit is set, memory block is
-				allocated */
-};
-/*
- * Memory pool mechanism uses one array of k_mem_pool_quad_block for accounting
- * blocks of one size. Block sizes go from maximal to minimal. Next memory
- * block size is 4 times less than the previous one and thus requires 4 times
- * bigger array of k_mem_pool_quad_block structures to keep track of the
- * memory blocks.
- */
-
-/*
- * The array of k_mem_pool_block_set keeps the information of each array of
- * k_mem_pool_quad_block structures
- */
-struct k_mem_pool_block_set {
-	size_t block_size; /* memory block size */
-	u32_t nr_of_entries; /* nr of quad block structures in the array */
-	struct k_mem_pool_quad_block *quad_block;
-	int count;
+struct k_mem_pool_lvl {
+	union {
+		u32_t *bits_p;
+		u32_t bits;
+	};
+	sys_dlist_t free_list;
 };
 
-/* Memory pool descriptor */
 struct k_mem_pool {
-	size_t max_block_size;
-	size_t min_block_size;
-	u32_t nr_of_maxblocks;
-	u32_t nr_of_block_sets;
-	struct k_mem_pool_block_set *block_set;
-	char *bufblock;
+	void *buf;
+	size_t max_sz;
+	u16_t n_max;
+	u8_t n_levels;
+	u8_t max_inline_level;
+	struct k_mem_pool_lvl *levels;
```

Comment:
Yeah this looks ok to me, I don't see any place where padding would be inserted..